packages/require/lib/types.ts

```diff
@@ -0,0 +1,112 @@
+import TruffleConfig from "@truffle/config";
+import Resolver from "@truffle/resolver";
+import TruffleError from "@truffle/error";
+
+export interface RequireOptions {
+  /**
+   * The path to the source file to be required/executed
+   */
+  file: string;
+
+  /**
+   * The `@truffle/config` for the user's project
+   */
+  config: TruffleConfig;
+}
+
+export interface ExecOptions {
+  contracts_build_directory: string;
+  /**
+   * Path to file to execute. Must be a module that exports a function.
+   */
+  file: string;
+
+  /**
+   * Object containing any global variables you'd like set when this function is
+   * run.
+   */
+  context: {
+    [index: string]: any;
+  };
+
+  /**
+   * The instance of `@truffle/resolver` to be passed into the global context
+   * when executing the script
+   */
+  resolver: Resolver;
+
+  /**
+   * The JSON RPC provider that will be used by the interface-adpater and web3
+   * instance injected into the script
+   */
+  provider: Provider;
+
+  /**
+   * A reference to a network defined in the user's `@truffle/config`.
+   */
+  network: string | number;
+
+  /**
+   * The id of the network as reported by the `net_version` JSON RPC method.
+   */
+  network_id: string | number;
+
+  /**
+   * The configuration for any networks that may be referenced by the script
+   * being executed.
+   */
+  networks: {
+    [index: string]: {
+      type: string; // likely "ethereum", but not guaranteed
+    };
+  };
+}
+
+/**
+ * The body of a valid JSON RPC request
+ */
+interface JsonRpcRequest {
+  jsonrpc: string;
+  method: string;
+  params: any[];
+  id: number;
+}
+
+/**
+ * The body of a valid JSON RPC response
+ */
+export interface JsonRpcResponse {
+  jsonrpc: string;
+  id: number;
```

Comment:
I think that while this is a more permissive type restriction for this field, it would actually increase the degree of restriction over the objects that can be assigned to this interface. As I mentioned in my other comment, these types need to be assignable from web3 providers, so I don't think I can further restrict them beyond what's already here.

That said, I do think it would be beneficial for Truffle to formally define a provider type (and subsequent subtypes, such as this one) along with a set of wrapper/transformer/converter/etc types/utilities/proxies/whatevers that allow Truffle to accept providers from various libraries, while still presenting them with a consistent abstraction internally. I just don't think that doing that is in scope for this issue.

Reply:
I'm not following your reasoning.

JSON-RPC 2.0 spec: https://www.jsonrpc.org/specification#request_object

Here is the type in web3: https://github.com/ChainSafe/web3.js/blob/1.x/packages/web3-core-helpers/types/index.d.ts#L225

Reply:
Yeah, you're totally right, and I definitely had my wires crossed yesterday when writing that.

But also I swear I copied these types directly out of the web3 code for the version that's currently in Truffle's `yarn.lock`. I wonder if they modified it since then?

Regardless as long as we're not making it so that providers from earlier versions are not assignable to variables/args that are typed with our provider type, I'm happy.